Spigot-API-Patches/0132-Custom-plugin-advancement-API.patch

```diff
@@ -0,0 +1,260 @@
++    private final AdvancementRewards rewards;
++    private final NamespacedKey key;
++
++    public AdvancementData(@Nullable NamespacedKey parent, AdvancementDisplay display, AdvancementRewards rewards, NamespacedKey key) {
++        this.parent = parent;
++        this.display = display;
++        this.rewards = rewards;
++        this.key = key;
++    }
++
++    public NamespacedKey getKey() {
++        return key;
++    }
++
++    /**
++     * The optional parent advancement of this advancement (does not apply for the root advancement).
++     * @return The parent advancement if present otherwise null
++     */
++    @Nullable
++    public NamespacedKey getParent() {
++        return parent;
++    }
++
++    /**
++     * The optional display data. Controls how the advancement looks for the client.
++     * @return Display data if present otherwise null
++     */
++    @Nullable
++    public AdvancementDisplay getDisplay() {
++        return display;
++    }
++
++    /**
++     * Required rewards object which gives rewards when a player completes the advancement.
++     * Can be used to define a custom function for executing commands (which can then execute code..)
++     * @return The rewards
++     */
++    public AdvancementRewards getRewards() {
++        return rewards;
++    }
++
++    public static class AdvancementRewards {
++        private int experience;
++        private NamespacedKey[] loot;
++        private NamespacedKey[] recipes;
++        @Nullable
++        private NamespacedKey customFunction;
++
++        public AdvancementRewards(int experience) {
++            this(experience, new NamespacedKey[0], new NamespacedKey[0], null);
++        }
++
++        public AdvancementRewards(int experience, NamespacedKey customFunction) {
++            this(experience, new NamespacedKey[0], new NamespacedKey[0], customFunction);
++        }
++
++        public AdvancementRewards(NamespacedKey customFunction) {
++            this(0, new NamespacedKey[0], new NamespacedKey[0], customFunction);
++        }
++
++        public AdvancementRewards(int experience, NamespacedKey[] loot, NamespacedKey[] recipes, @Nullable NamespacedKey customFunction) {
++            this.experience = experience;
++            this.loot = loot;
++            this.recipes = recipes;
++            this.customFunction = customFunction;
++        }
++
++        public int getExperience() {
++            return experience;
++        }
++
++        public NamespacedKey[] getLoot() {
++            return loot;
++        }
++
++        public NamespacedKey[] getRecipes() {
++            return recipes;
++        }
++
++        /**
++         * Functions are text files with the extension .mcfunction in the datapacks folder of your main world
++         */
++        @Nullable
++        public NamespacedKey getCustomFunction() {
++            return customFunction;
++        }
++    }
++
++    public class AdvancementDisplay {
```

Comment:
static class

Reply:
Does it need to be? Compiles already :confused:

Reply:
if it has no reference to the parents instance, it should be, otherwise they are harder linked.

Reply:
actually lets just extract all these inner classes and enums anyways into the new paper package space

org.destroystokyo.paper.advancements

Reply:
ok

Reply:
Was thinking this myself 👍